providers/azure.go

```diff
@@ -233,11 +236,35 @@ func getEmailFromJSON(json *simplejson.Json) (string, error) {
 	return email, err
 }
 
+func getEmailFromIDToken(idToken string) (string, error) {
+	jwt := strings.Split(idToken, ".")
+	jwtData := strings.TrimSuffix(jwt[1], "=")
+	b, err := base64.RawURLEncoding.DecodeString(jwtData)
+	if err != nil {
+		return "", fmt.Errorf("jwt is malformed: %w", err)
+	}
+
+	c := struct {
+		Email string `json:"email"`
+	}{}
+
+	err = json.Unmarshal(b, &c)
+	if err != nil {
+		return "", fmt.Errorf("unable to unmarshal jwt payload: %w", err)
+	}
+	if c.Email == "" {
+		return "", errors.New("missing email claim from id_token")
+	}
+	return c.Email, nil
+}
+
 // GetEmailAddress returns the Account email address
 func (p *AzureProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
 	var email string
-	var err error
 
+	if s.IDToken != "" {
+		return getEmailFromIDToken(s.IDToken)
+	}
```

Comment:
It would be better to first try the ID Token extraction, and then try the profile URL if the ID Token does not include an email.

Out of interest, comparing this with the OIDC provider, I noticed that the Azure provider doesn't extract User/Email/Groups during the Redeem process, but the OIDC provider does. Might be better to try and bring the functionality inline with the OIDC provider and try to extract claims from the ID Token during redeem, and then leave `GetEmailAddress` as a fallback in case the ID Token doesn't include the email

Reply:
FYI - `GetEmailAddress` should be considered deprecated in favor of `EnrichSessionState`.

`EnrichSessionState` is called immediately after `redeem` logic in `OAuthCallback`. It is now the place to do any getters & setters on the session.

Historically, a lot of people hacked a lot of unrelated functionality into `GetuserName` & `GetEmailAddress` because those were the only post `redeem` hooks to hit things like the ProfileURL. Now `EnrichSessionState` and `Authorize` are the appropriate methods for various post redeem logic.

Reply:
For more context: https://github.com/oauth2-proxy/oauth2-proxy/issues/883

Reply:
@NickMeves doesn't [enrichSessionState()](https://github.com/oauth2-proxy/oauth2-proxy/blob/master/oauthproxy.go#L410) call `GetEmailAddress()` first. Do you suggest remove `GetEmailAddress()` from Azure provider?

Reply:
The intention is all previous uses of `GetUserName` and `GetEmailAddress` (where the original intent was they were just getters, but a bazillion providers made those methods mutate the session) make the switch to `EnrichSessionState`.

I've refactored all users of `GetUserName` already so it has been removed. Trying to hit `GetEmailAddress` as people touch it in PRs.

You probably just need to rename the method and tweak the logic a little bit to get the returned email into the `session.Email`.

Reply:
@NickMeves i've made the refactoring, PTAL
@JoelSpeed I made the change to extract email from id token as best effort. Regarding the feedback to move email extraction into `Redeem()` logic, I think it's still arguable as I see different providers all implement it differently. For my naive understanding, `Redeem` should just redeem for the token. How to interpret and process these tokens seems to be EnrichSessionState's business?

Reply:
I think any code-ful extraction of an IDToken claim could be in Redeem or EnrichSessionState, both work. I think you'll see most other providers handle this in Redeem if they have it available to them. But that might be a legacy byproduct since `GetUserName` and `GetEmailAddress` were very tailored to a network call to retrieve these details from a profile URL or something.

If there's any network traffic, that 100% belongs in `EnrichSessionState`

Reply:
Added context: https://github.com/oauth2-proxy/oauth2-proxy/issues/883

Reply:
@NickMeves can we refactor that in a separate PR? :)

Reply:
Doing a quick eye test, you seem fine doing all the logic in `EnrichSessionState` - I don't see what you need to refactor

Reply:
let me know what else I need to do. thanks!

Reply:
> It would be better to first try the ID Token extraction, and then try the profile URL if the ID Token does not include an email.
> 
> Out of interest, comparing this with the OIDC provider, I noticed that the Azure provider doesn't extract User/Email/Groups during the Redeem process, but the OIDC provider does. Might be better to try and bring the functionality inline with the OIDC provider and try to extract claims from the ID Token during redeem, and then leave `GetEmailAddress` as a fallback in case the ID Token doesn't include the email

So upon closer look at the PR, I would actually strongly agree with @JoelSpeed's point -- we should really aim to consolidate the logic for claim extration from an IDToken that works across providers.  Here's the logic from OIDC: https://github.com/oauth2-proxy/oauth2-proxy/blob/8bed7aafbdb2cb6241b9655e983dd54d54b50531/providers/oidc.go#L229

(Ignore the requests to the profileURL as a fallback, I intend to move that to `EnrichSessionState` in OIDC: https://github.com/oauth2-proxy/oauth2-proxy/issues/883).

Big reason we should consolidate: A frequent feature request we intend to implement soon is arbitrary claim -> header mappings, so users have full control of passing claims to their backend apps in `X-Forwarded-*` headers. If any provider that has IDTokens with claims uses a common function, adding that functionality will apply to much more providers than just OIDC.

Reply:
Also, what support does Azure provide for OIDC discovery documents (`openid-configuration` & `jwks_uri`) that an `oidc.IDTokenVerifier` would use?

The reason I ask is if Azure is mostly OIDC ready, we can also make it support Bearer auth. I have this refactor in flight as well that aims to make any provider that supports OIDC Verifiers be able to be used with bearer token auth:
https://github.com/oauth2-proxy/oauth2-proxy/pull/869#discussion_r527412285

Reply:
I will work on moving the email extraction to `Redeem`

> Also, since I see the Azure provider gives back an ID Token, what support does Azure provide for OIDC discovery documents (openid-configuration & jwks_uri) that an oidc.IDTokenVerifier would use?

when you specify the AAD issuer, oidc library will be able to retrieve the oidc discovery document.
https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-protocols-oidc#fetch-the-openid-connect-metadata-document

Reply:
To be clear, it is Okay for `Redeem` to make network call to download jwks to verify the id_token, right?

Reply:
> To be clear, it is Okay for `Redeem` to make additional network call to download jwks to verify the id_token, and it is expected for each provider to implement it, right?

FWIW, I only see oidc and gitlab providers doing token verification in `Redeem`. Let me know what is the desired pattern going forward.

Reply:
Do it in `EnrichSessionState`.

And it is uniform, refactor the claim extraction logic to a helper method in `util.go` that this provider & OIDC consume. This way future improvements to it will be available in Azure for you and not just OIDC.

If that isn't an easy refactor, there's also this sample of a cleaner way to get JWT claims than the current logic: https://github.com/oauth2-proxy/oauth2-proxy/blob/8bed7aafbdb2cb6241b9655e983dd54d54b50531/providers/logingov.go#L104

Reply:
i left the token claim extraction in `EnrichSessionState` and did a small refactoring to use `jwt` package to extract claims. Hope it would suffice.